src/coreclr/vm/proftoeeinterfaceimpl.cpp

```diff
@@ -7615,6 +7619,95 @@ HRESULT ProfToEEInterfaceImpl::GetNonGCHeapBounds(ULONG cObjectRanges,
     return S_OK;
 }
 
+HRESULT ProfToEEInterfaceImpl::EnumerateGCHeapObjects(ObjectCallback callback, void* callbackState)
+{
+    CONTRACTL
+    {
+        // Yay!
+        NOTHROW;
+
+        // Suspending EE is contracted to trigger GC through ThreadSuspend::SuspendAllThreads
+        GC_TRIGGERS;
+
+        // Yay!
+        MODE_ANY;
+
+        // Suspending EE will acquire the ThreadStore Lock via ThreadSuspend::LockThreadStore.
+        CAN_TAKE_LOCK;
+
+        // Yay!
+        EE_THREAD_NOT_REQUIRED;
+    }
+    CONTRACTL_END;
+
+    PROFILER_TO_CLR_ENTRYPOINT_ASYNC_EX(kP2EEAllowableAfterAttach | kP2EETriggers,
```

Comment:
I noticed, somewhat surprisingly, that the implementation of the ASYNC macro doesn't do any validation based on the kP2EETriggers flag. While we might want to improve the macro itself in the future, given we are near the end of feature development and its not causing known problems its probably safer to add a comment and some additional validation local to this method. Something like:

```
PROFILER_TO_CLR_ENTRYPOINT_ASYNC_EX(...);
if(!IsCalledAsynchronously() && !AreCallbackStateFlagsSet(COR_PRF_CALLBACKSTATE_IN_TRIGGERS_SCOPE))
{
    LOG((LF_CORPROF, LL_ERROR,
                 "**PROF: ERROR: Returning CORPROF_E_UNSUPPORTED_CALL_SEQUENCE "
                 "due to illegal asynchronous profiler call\n"));
    return CORPROF_E_UNSUPPORTED_CALL_SEQUENCE;
}
```

This should catch most misuse though it wouldn't detect complicated cases like a profiler that hijacks a thread running managed code and then tries to do enumeration from inside the hijack. I'm sure we could make a stronger check if we needed to, but I think this would be good for now.

Reply:
Good catch, added the extra check to SuspendRuntime, ResumeRuntime, and EnumerateGCHeapObjects